### PR TITLE
Feature: Add fuzzy single select prompt

### PR DESCRIPTION
**Pull request by deg0nz**
This superseeds #79.

I added all the changes that where requested by @pksunkara in #79.

I added a cursor simulation, so we have a basic functionality to change the search term conveniently. Currently, the cursor is fixed to be black font on white background, but I'd like this thing to be possible to be styled by the user. 

Here's how it currently looks:

fuzzy_cursor

This PR is to be discussed. Maybe I forgot something and some stuff is still to do (see below).

**This stuff is still to be done until I consider this PR "ready to merge":**

- [x] Make cursor styleable (or better: detect terminal colors and style it based on that)
- [x] Remove/change old "paging code" and assume the fuzzy select narrows it down to 1 page (as commented by @pksunkara in #79)

### CODE REVIEW

**Comment by deg0nz** [comment]
I would like to detect the current terminal colors and style the cursor based on that.

@pksunkara do you know, if this is possible to find out via the `console` crate?

File: src/lib.rs [hunk 1]
```diff
@@ -27,6 +27,9 @@ pub use prompts::{
 pub use validate::Validator;
 
 #[cfg(feature = "editor")]
```
Comment (pksunkara) [inline src/lib.rs:29]: This is for `mod edit` IIRC.

File: src/prompts/fuzzy_select.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    paged: bool,
+    offset: usize,
+    lines_per_item: usize,
+    fuzzy_search_is_case_sensitive: bool,
+}
+
+impl<'a> FuzzySelect<'a> {
+    /// Creates the prompt with a specific text.
+    pub fn new() -> Select<'static> {
+        Select::with_theme(&SimpleTheme)
```
Comment (pksunkara) [inline src/prompts/fuzzy_select.rs:55]: `Select`?

File: src/theme.rs [hunk 3]
```diff
@@ -203,6 +203,51 @@ pub trait Theme {
...
+        prompt: &str,
+        search_term: &str,
+        cursor_pos: usize
+    ) -> fmt::Result {
+
+        if prompt.is_empty() {
+            write!(
+                f,
+                "> ",
+            )?;
```
Comment (pksunkara) [inline src/theme.rs:221]: You don't need this statement.

File: src/theme.rs [hunk 4]
```diff
@@ -203,6 +203,51 @@ pub trait Theme {
...
+                f,
+                "{} ",
+                prompt,
+            )?;
+        }
+
+        if cursor_pos < search_term.len() {
+            let st_head = search_term[0..cursor_pos].to_string();
+            let st_tail = search_term[cursor_pos+1..search_term.len()].to_string();
+            let st_cursor = style(search_term.to_string().chars().nth(cursor_pos).unwrap()).black().on_white();
```
Comment (pksunkara) [inline src/theme.rs:233]: No `style` here

File: src/theme.rs [hunk 5]
```diff
@@ -203,6 +203,51 @@ pub trait Theme {
...
+            let st_cursor = style(search_term.to_string().chars().nth(cursor_pos).unwrap()).black().on_white();  
+            write!(
+                f,
+                "{}{}{}",
+                st_head,
+                st_cursor,
+                st_tail
+            )  
+        } else {
+            let cursor = style(" ".to_string()).black().on_white();
```
Comment (pksunkara) [inline src/theme.rs:242]: No `style` here

File: src/prompts/fuzzy_select.rs [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+            {
+                render.select_prompt_item(item, idx == sel)?;
+                term.flush()?
+            }
+
+            match term.read_key()? {
+                Key::Escape if allow_quit => {
+                    if self.clear {
+                        term.clear_last_lines(filtered_list.len())?;
+                        term.flush()?
```
Comment (pksunkara) [inline src/prompts/fuzzy_select.rs:222]:
```suggestion
                        term.flush()?;

```

File: src/prompts/fuzzy_select.rs [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+pub struct FuzzySelect<'a> {
+    default: usize,
+    items: Vec<String>,
+    prompt: String,
+    clear: bool,
+    theme: &'a dyn Theme,
+    paged: bool,
+    offset: usize,
+    lines_per_item: usize,
+    fuzzy_search_is_case_sensitive: bool,
```
Comment (pksunkara) [inline src/prompts/fuzzy_select.rs:49]:
```suggestion
    ignore_casing: bool,
```

**Review by pksunkara — CHANGES_REQUESTED** [review]

File: src/prompts/fuzzy_select.rs (hunk 2 above)
Comment (deg0nz) [reply src/prompts/fuzzy_select.rs]: Fixed. Don't know how that got in there 😅

File: src/lib.rs (hunk 1 above)
Comment (deg0nz) [reply src/lib.rs]: Fixed.

File: src/theme.rs (hunk 3 above)
Comment (deg0nz) [reply src/theme.rs]: Removed.

File: src/prompts/fuzzy_select.rs (hunk 6 above)
Comment (deg0nz) [reply src/prompts/fuzzy_select.rs]: Fixed.

File: src/prompts/fuzzy_select.rs (hunk 7 above)
Comment (deg0nz) [reply src/prompts/fuzzy_select.rs]: Done.

**Comment by deg0nz** [comment]
@pksunkara after your suggestion to change the `ignore_casing` variable, I stumbled upon the fact, that this variable changes nothing regarding the fuzzy searcher itself. It just adds the characters as lower-case to the search string.

I got nosy and looked into the code of the `fuzzy-matcher` code. The `Default` implementation of `SkimMatcherV2` already [uses "smart" case sensitive matching](https://github.com/lotabout/fuzzy-matcher/blob/ee73fa4559a9ea06e51b54b0a436590c911c18a1/src/skim.rs#L621) which gives penalty if the case is not matched and bonus if the case is matched.

I suggest, **we remove the whole case-sensitive thing**, since the `fuzzy-matcher` already handles that.

**Comment by pksunkara** [comment]
Sounds good to me.

File: examples/fuzzyselect.rs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+fn main() {
+    let selections = &[
+        "Ice Cream",
+        "Vanilla Cupcake",
+        "Chocolate Muffin",
+        "A Pile of sweet, sweet mustard",
+    ];
+
+    let position = FuzzySelect::with_theme(&ColorfulTheme::default())
```
Comment (pksunkara) [inline examples/fuzzyselect.rs:11]:
```suggestion
    let selection = FuzzySelect::with_theme(&ColorfulTheme::default())
```

File: src/prompts/fuzzy_select.rs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            lines_per_item: 1,
+            ignore_casing: false,
+        }
+    }
+
+    /// Enables or disables paging
+    pub fn paged(&mut self, val: bool) -> &mut FuzzySelect<'a> {
+        self.paged = val;
+        self
+    }
```
Comment (pksunkara) [inline src/prompts/fuzzy_select.rs:74]: Let's remove paging completely for fuzzy select

**Review by pksunkara — CHANGES_REQUESTED** [review]

File: examples/fuzzyselect.rs (hunk 8 above)
Comment (deg0nz) [reply examples/fuzzyselect.rs]: done

**Review by pksunkara — APPROVED** [review]

**Comment by Araxeus** [comment]
@pksunkara why is there no documentation for this feature?

Did I somehow miss it?

**Comment by pksunkara** [comment]
Release 0.10.1 which should have documentation about it.

**Pull request merged**